src/value.rs

```diff
@@ -0,0 +1,170 @@
+use pyo3::{Bound, IntoPyObject, PyAny, PyErr, Python};
+use pyo3::prelude::PyDictMethods;
+use pyo3::types::{PyBool, PyDict, PyInt, PyString};
+use scylla_cql::_macro_internal::{ColumnType, DeserializationError, DeserializeValue, FrameSlice, UdtIterator};
+use scylla_cql::deserialize::value::{deser_error_replace_rust_name, FixedLengthBytesSequenceIterator};
+use scylla_cql::frame::response::result::{CollectionType, NativeType};
+use scylla_cql::frame::types;
+
+// NOTE: I intentionally do NOT use Scylla's `DeserializeValue` trait here.
+// The trait does not provide a `Python` argument, meaning that Python objects which
+// would have to be constructed inside `deserialize()` or deeper in recursion
+// would require acquiring the GIL separately for every column/element/UDT/List
+// during nested decoding.
+//
+// Using our own `PyDeserializeValue` trait keeps all Python conversions inside
+// a single `Python::with_gil(...)` boundary, allowing complex values (lists,
+// sets, UDTs, nested collections) to be deserialized without repeated GIL
+// acquisition. This avoids potential slowdown of acquiring GIL multiples times.
+//
```

Comment:
Note: We wouldn't neet to acquire and release GIL for each value.
We would need to acquire it once, and then in each value check that it is already acquired.

https://pyo3.rs/v0.27.1/performance.html#access-to-bound-implies-access-to-python-token

So the reasoning is fine, but the comment is a bit imprecise.